### ISSUE TITLE
Only "getter" methods should return Optional<> since java.util.Optional is not serializable

### ISSUE THREAD

**Issue by tskardal** [immutables/immutables#240]
I just started exploring this library, and it looks promising. However, I noticed that fields that are specified as Optional<T> are stored as Optional<T> in the generated code instead of just exposing them as optional. This was brought to my attention the other day when I read the ZeroTurnaround Java 8 best practices [cheat sheet](http://zeroturnaround.com/rebellabs/java-8-best-practices-cheat-sheet/).

More on this in this [SO question](http://stackoverflow.com/a/24564612/48431).

Example

```java
@Value.Immutable
public interface Person {
  Optional<String> name();
}
```
becomes
```java
public final class ImmutablePerson implements Person {
  private final Optional<String> name;
  // ...
  public Optional<String> name() {
    return name;
  }
}
```
but should probably be
```java
public final class ImmutablePerson implements Person {
  private final String name;
  // ...
  public Optional<String> name() {
    return Optional.ofNullable(name);
  }
}
```

**Comment by elucash** [comment]
Thank you for the feature request! All in all this may be a doubtful practice, but I, guess, we should use `java.util.Optional*` as they are "designed", keeping in mind that we have alternative of using Guava or Fugue variants which remain boxed in _Immutables_ as before. The experimental implementation has landed, but it requires some more testing before releasing.

**Comment by tskardal** [comment]
Cool :) I didn't know this before reading the linked posts. As I see it the only reason to not declare fields as Optionals are the fact that they aren't serializable. IIRC idiomatic Scala defines optional fields without a default value as `Option`s

**Comment by elucash** [comment]
released in 2.1.9

**Comment by jbergfi** [comment]
I've been using Immutables 2.0.17 for quite some time now and serializing my value objects using Kryo. I'm now looking at updating Immutables to latest version to get support for Eclipse Neon but struck into regression problems due to this change.

I must be able to support serialization compatibility with the code running 2.0.17 but now, when running the latest version and deserializing an optional, Kryo is expecting the object directly instead of an Optional and when it tries to deserialize an object from 2.0.17 it fails to deserialize properly.

I'm trying to figure out a workaround but haven't yet managed to. Would it be possible to add a flag to say if the optional should be unboxed or not? Another option would be to get Eclipse Neon support for 2.0.X I guess but I would rather update to the latest to get all the latest features.

Any ideas?

**Comment by elucash** [comment]
I don't think we need to add such a flag specifically for this case. Binary serialization without explicit schema should not be ever consider for long term storage, so it's usually not expected to survive version changes. Try to read bytes using old version of types then convert to some neutral representation, like JSON, then read it with newer version. To make migration easier you can try to switch to using `@Serial.Structural` serialization which always uses nulls for optionals and tell Kryo to respect java serialization rules like `readResolve` and `writeReplace` if there's such option. More heavyweight, but working solution might be using new experimental "encoding" functionality to change the way `java.util.Optional` is implemented, please see #363. Additional hint - `java.util.Optional` is not serializable, so it's implemented as nullable field, but other kinds of optional types such as the one from Guava are still stored as optional fields, so you can use those as an alternative.

**Comment by jbergfi** [comment]
I do agree JSON is much better for long term persistent storage and we do that quite a bit but unfortunately we have a few places still that store kryo/java serialized objects as well persistently. Even if that wouldn't be the case I still need backwards compatibility for the serialization in my multi-node system so I can update one node at a time while keeping it fully functional during the update.

Earlier we used the @Serial.Structural which works well but is a bit slow which is the reason we recently switched to using CompatibleFieldSerializer of Kryo which essentially does the same thing but faster and with less bytes. On the downside it's dependent quite heavily on what exactly gets generated in the class.

Anyhow, I've managed to create a workaround now by extending Kryos CompatibleFieldSerializer and highjacking unboxed java.util.Optionals by looking if there's a method with the same name as the field which returns an Optional. For these fields I make sure they are serialized just like if they would be boxed as Optionals in the first place and corresponding tricks for deserialize. Good thing about this is that I can add a new serialization format to my system which doesn't at all need to bother with Optionals which saves a few bytes and should be slightly faster while I still can support the old serialization format.

So yeah, it seems I can manage without the flag, thanks anyway!
